pkg/resourcemanager/controller/node/reconciler.go

```diff
@@ -57,7 +67,115 @@ func (r *Reconciler) Reconcile(reconcileCtx context.Context, req reconcile.Reque
+func AllNodeCriticalDaemonPodsAreScheduled(log logr.Logger, recorder record.EventRecorder, node *corev1.Node, daemonSets []appsv1.DaemonSet, nodeCriticalPods []corev1.Pod) bool {
+	// collect a set of all scheduled DaemonSets on the node
+	scheduledDaemonSets := sets.NewString()
+	for _, pod := range nodeCriticalPods {
+		controllerRef := metav1.GetControllerOf(&pod)
+		if controllerRef == nil || schema.FromAPIVersionAndKind(controllerRef.APIVersion, controllerRef.Kind) != daemonSetGVK {
+			continue
+		}
+
+		scheduledDaemonSets.Insert(string(controllerRef.UID))
+	}
+
+	// filter for DaemonSets that were not scheduled to the node yet
+	var unscheduledDaemonSets []client.ObjectKey
+	for _, daemonSet := range daemonSets {
+		if daemonSet.Spec.Template.ObjectMeta.Labels[v1beta1constants.LabelNodeCriticalComponent] != "true" {
+			continue
+		}
+
+		// determine whether DaemonSet needs to be scheduled to the node at all
+		if shouldRun, _ := helper.NodeShouldRunDaemonPod(node, &daemonSet); !shouldRun {
+			continue
+		}
+
+		// check whether DaemonSet has corresponding daemon pod on the node
+		key := client.ObjectKeyFromObject(&daemonSet)
+		if !scheduledDaemonSets.Has(string(daemonSet.UID)) {
+			unscheduledDaemonSets = append(unscheduledDaemonSets, key)
+		}
+	}
+
+	if len(unscheduledDaemonSets) > 0 {
+		log.Info("Node-critical DaemonSets found that were not scheduled to Node yet", "daemonSets", unscheduledDaemonSets)
+		recorder.Eventf(node, corev1.EventTypeWarning, "UnscheduledNodeCriticalDaemonSets", "Node-critical DaemonSets found that were not scheduled to Node yet: %s", objectKeysToString(unscheduledDaemonSets))
+		return false
+	}
+
+	return true
+}
+
+// AllNodeCriticalPodsAreReady returns true if all the given pods are ready by checking their Ready conditions.
+func AllNodeCriticalPodsAreReady(log logr.Logger, recorder record.EventRecorder, node *corev1.Node, nodeCriticalPods []corev1.Pod) bool {
+	var unreadyPods []client.ObjectKey
+	for _, pod := range nodeCriticalPods {
+		if !health.IsPodReady(&pod) {
+			unreadyPods = append(unreadyPods, client.ObjectKeyFromObject(&pod))
+		}
+	}
+
+	if len(unreadyPods) > 0 {
+		log.Info("Unready node-critical Pods found on Node", "pods", unreadyPods)
+		recorder.Eventf(node, corev1.EventTypeWarning, "UnreadyNodeCriticalPods", "Unready node-critical Pods found on Node: %s", objectKeysToString(unreadyPods))
+		return false
+	}
+
+	return true
+}
+
+// RemoveTaint removes the taint managed by this controller from the given node object
+func RemoveTaint(ctx context.Context, w client.Writer, node *corev1.Node) error {
+	patch := client.MergeFromWithOptions(node.DeepCopy(), client.MergeFromWithOptimisticLock{})
+	taints := node.Spec.Taints
+	for i := 0; i < len(taints); i++ {
+		if taints[i].Key == v1beta1constants.TaintNodeCriticalComponentsNotReady {
+			taints = append(taints[:i], taints[i+1:]...)
+			i--
+		}
+	}
+	node.Spec.Taints = taints
```

Comment:
This might work, but isn't it artificially complicated? 😅 What about just creating a new slice with the desired taints?